### PR TITLE
Hackmon header

### PR DESCRIPTION
**Pull request by alexgirard**
### Summary
- sorry my branch name sucks but it has both the headers and the implementation for the `Hackmon`, `Player`, and `Species` classes

### CODE REVIEW

File: game/Makefile [hunk 1]
```diff
@@ -1,7 +1,7 @@
 CXX = g++
 CXXFLAGS = -std=c++14 -Wall -MMD
 EXEC = hackmon
-OBJECTS = action.o
```
Comment (BenLanglois) [inline game/Makefile:4]: Wow you removed action.o :(
bitmoji

File: game/hackmon.cc [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#include <vector>
+#include <string>
```
Comment (BenLanglois) [inline game/hackmon.cc:2]: Looks like you don't need string or map anymore

File: game/hackmon.cc [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+#include <memory>
+
+#include "hackmon.h"
+#include "debuffName.h"
+#include "statInfo.h"
+using namespace std;
+
+Hackmon::Hackmon(string hackmonName, vector<Move> moves, Stats stats, vector<Family> family):
+name{hackmonName}, moves{moves}, stats{stats}, family{family} {
+  currentMove = nullptr;
```
Comment (BenLanglois) [inline game/hackmon.cc]: `currentMove`, `inBattle`, and `debuff` should be set in the MIL. Also you're missing `mobility`.

File: game/hackmon.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+#include "move.h"
+#include "stats.h"
+#include "family.h"
+#include "mobility.h"
+
+class Hackmon {
+ public:
+  std::string name;
+  Move& currentMove;
```
Comment (BenLanglois) [inline game/hackmon.h]: This should be a pointer, not a reference

File: game/hackmon.cc (hunk 2 above)
Comment (alexgirard) [reply game/hackmon.cc:2]: why wouldn't I need string?

**Review by ariellam — COMMENTED** [review]
lgtm 👍

File: game/hackmon.h [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  std::unique_ptr<Debuff> debuff;
+  Mobility mobility;
```
Comment (BenLanglois) [inline game/hackmon.h]: Mobility is an abstract base type and can't be statically allocated. This should be a unique pointer

File: game/hackmon.cc (hunk 2 above)
Comment (BenLanglois) [reply game/hackmon.cc:2]: Never mind, I didn't see `hackmonName`

File: game/hackmon.h [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ public:
+  std::string name;
+  std::vector<Move> moves;
+  Stats stats;
+  std::vector<Family> family;
+  Mobility mobility;
+  Move& currentMove;
+  std::unique_ptr<Debuff> debuff;
+
+  Hackmon(std::string, std::vector<Move>, Stats, std::vector<Family>, Mobility);
```
Comment (BenLanglois) [inline game/hackmon.h]: Same thing here, you can't have a parameter of type Mobility because it's abstract

File: game/player.h [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+#include <memory>
+#include <utility>
+
+#include "hackmon.h"
+#include "statInfo.h"
+
+class Player {
+ public:
+  std::string name;
+  std::vector<unique_ptr<Hackmon>> party;
```
Comment (BenLanglois) [inline game/player.h]: missing `std::` for the `unique_ptr`

File: game/player.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  int winTotal;
+  int nextAlive;
+
+  Player(std::string, std::vector<unique_ptr<Hackmon>>);
```
Comment (BenLanglois) [inline game/player.h]: Also missing `std::` here

File: game/player.cc [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+#include <memory>
+#include <utility>
+
+#include "player.h"
+#include "hackmon.h"
+#include "statInfo.h"
+using namespace std;
+
+Player::Player(string name, vector<unique_ptr<Hackmon>> party): name{name}, party{party}, winTotal{0} {
+  nextAlive = numberBattling; // FIXME: global variable
```
Comment (BenLanglois) [inline game/player.cc]: This should probably be done in the MIL

File: game/player.cc [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+  for (auto& index : numberBattling) {
```
Comment (BenLanglois) [inline game/player.cc]: You can't iterate over `numberBattling` because it's an int. Probably easiest to just use a regular for loop

Comment (BenLanglois) [reply game/player.cc]:
```suggestion
  for (size_t index = 0; index < numberBattling; ++index) {
```

File: game/player.h [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+#include "hackmon.h"
+#include "statInfo.h"
+
+class Player {
+ public:
+  std::string name;
+  std::vector<unique_ptr<Hackmon>> party;
+  int winTotal;
+  int nextAlive;
```
Comment (BenLanglois) [inline game/player.h]: `winTotal` and `nextAlive` should probably be unsigned

Comment (BenLanglois) [reply game/player.h]: This is also true for the parameters to `isAlive`, `getHackmon`, and `swapHackmon`

File: game/player.h (hunk 7 above)
Comment (BenLanglois) [inline game/player.h]: It looks like you never declared `inBattle`

File: game/player.cc [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+// move calls this function -- returns nullptr if no hackmon left
+Player::Hackmon* getHackmon(int index) {
+  if (isAlive(*(inBattle.at(index)))) {
+    return party.at(*(inBattle.at(index)));
+  }
+  else {
+    if (nextAlive < party.size()) {
+      inBattle.at(index) = *party.at(nextAlive++);
```
Comment (BenLanglois) [inline game/player.cc]: inBattle doesn't point to the actual Hackmon object, but rather the `unique_ptr` to a Hackmon object. You don't need the dereference here.
```suggestion
      inBattle.at(index) = party.at(nextAlive++);
```

File: game/player.cc [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+      return party.at(*(inBattle.at(index)));
```
Comment (BenLanglois) [inline game/player.cc]: This doesn't quite make sense. I don't think you need the call to `party.at`.
```suggestion
      return *(inBattle.at(index));
```

File: game/player.cc [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+  }
+}
+
+Player::bool isAlive(int index) {
+  return (*(party.at(index)).stats.getStat(HP) > 0);
+}
+
+// move calls this function -- returns nullptr if no hackmon left
+Player::Hackmon* getHackmon(int index) {
+  if (isAlive(*(inBattle.at(index)))) {
```
Comment (BenLanglois) [inline game/player.cc]: `isAlive` takes an int, but you're passing it a Hackmon ref

File: game/player.cc [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    return party.at(*(inBattle.at(index)));
```
Comment (BenLanglois) [inline game/player.cc]: Again, you don't need the call to `party.at` here (see my comment below)
```suggestion
    return *(inBattle.at(index));
```

File: game/species.h [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+#include <string>
+
+#include "hackmon.h"
+#include "move.h"
+#include "stats.h"
+#include "family.h"
+
+class Species {
+  std::string speciesName;
+  std::vector<Move> moves; // FIXME: should be moved to family? depends on type
```
Comment (BenLanglois) [inline game/species.h]: Move is abstract, you can't instantiate it

File: game/species.h [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+  Move specialMove;
+  Stats stats;
+  std::vector<Family> family;
+ public:
+  Species(const std::string, std::vector<Move>, Move, Stats, std::vector<Family>)
```
Comment (BenLanglois) [inline game/species.h]: Same here

**Review by BenLanglois — CHANGES_REQUESTED** [review]
Feel free to ignore most of my comments on player.cc since we're not using inbattle anymore.

File: game/player.h (hunk 7 above)
Comment (BenLanglois) [reply game/player.h]: As discussed, we are no longer implementing `Player` with `inBattle`.

File: game/hackmon.h [hunk 18]
```diff
@@ -13,14 +13,14 @@
 class Hackmon {
  public:
   std::string name;
-  std::vector<Move> moves;
+  std::vector<Move*> moves;
```
Comment (BenLanglois) [inline game/hackmon.h]: This should be a `vector<unique_ptr<Move>>` to follow RAII

File: game/player.cc [hunk 19]
```diff
@@ -7,37 +7,31 @@
...
-      inBattle.at(index) = nullptr;
       return nullptr;
     }
   }
 }
 
 // index1 from inBattle vector, index2 from party not in inBattle and not fainted
-Player::void swapHackmon(int index1, int index2) {
+Player::void swapHackmon(unsigned index1, unsigned index2) {
   swap(inBattle.at(index1), party.at(index2));
```
Comment (BenLanglois) [inline game/player.cc]: Looks like you forgot to remove `inBattle` here.

File: game/species.h [hunk 20]
```diff
@@ -11,14 +11,14 @@
 
 class Species {
   std::string speciesName;
-  std::vector<Move> moves; // FIXME: should be moved to family? depends on type
+  std::vector<Move*> moves; // FIXME: should be moved to family? depends on type
```
Comment (BenLanglois) [inline game/species.h]: This should also be a `unique_ptr`

**Review by BenLanglois — COMMENTED** [review]
Just a few small changes. Other than that LGTM.

**Review by ariellam — APPROVED** [review]
🚀

**Pull request merged**